### PR TITLE
http3: adding connect support

### PR DESCRIPTION
**Pull request by alyssawilk**
This includes validation for upgrade connects per Ryan's offline advice.
n.b. this should be a no-op for HTTP (where there is no mechanism to send both) and HTTP/2 (where nghttp2 validates) so not currently calling out in release notes.

Risk Level: low
Testing: new integration tests
Docs Changes: inline
Release Notes: n/a (quic alpha)

co-author: @DavidSchinazi

### CODE REVIEW

File: source/common/http/header_utility.cc
Comment (alyssawilk) [reply source/common/http/header_utility.cc:343]: scheme is default required for all HTTP/3 so I believe is validated elsewhere.
Path is required in the HCM unless it's CONNECT, but all envoy route matching asserts a path except for the special connect matchers, so Envoy will 404 unless you explicitly try to support CONNECT.

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:489]: yeah we could just land standard CONNECT support but I think folks are going to want websocket over h3 if they have it for H2.
I'll link to the H2 spec and your draft and tag it alpha. Alternately I can just add support for standard connect (block anything with path, regardless of protocol) and wait on the RFC.  thoughts?

File: source/common/http/header_utility.cc
Comment (alyssawilk) [reply source/common/http/header_utility.cc:344]: done

File: source/common/http/header_utility.cc
Comment (alyssawilk) [reply source/common/http/header_utility.cc:346]: yep, nice catch.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_server_stream.cc:160]: yeah CONNECT either has to have no path, or path + protocol.
The former is a standard CONNECT request, and will be blocked at the HCM unless there's CONNECT-specific route matchers (so off by default).  The latter gets transformed into an HTTP/1.1 style upgrade, and won't look like it's a CONNECT for the duration of the filter chain, will match normal routes so is disabled by default using a separate knob.

https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/upgrades
which now that I think about it I should update.

File: source/common/tcp_proxy/upstream.h
Comment (alyssawilk) [reply source/common/tcp_proxy/upstream.h:56]: correct.

File: test/config/utility.cc
Comment (alyssawilk) [reply test/config/utility.cc:675]: I think the H2 option is (sadly) misnamed because HTTP/2 allows standard connect without a special knob, and just disallows the upgrade style connect.

**Comment by alyssawilk** [comment]
Also I owe you sending that setting with HTTP/3 and how about a rename from upgrade_connect to extended_connect?

**Review by RyanTheOptimist — COMMENTED** [review]
> Also I owe you sending that setting with HTTP/3 and how about a rename from upgrade_connect to extended_connect?

Sounds perfect to me. That matches the IETF terminology for this H3 behavior and since this is an H3 protocol option, this seems like the right terminology.

**Comment by alyssawilk** [comment]
hm, I actually think I can't set custom settings.  @danzh2010 am I right in thinking this will need an upstream quiche change given FillSettingsFrame seems to be the way to change settings and doesn't have connect support?

**Comment by DavidSchinazi** [comment]
@alyssawilk yes that'll require a QUICHE change. I would suggest that if we change QUICHE to send that h3 setting, then QUICHE should also be validating that CONNECT requests carry the right pseudo-headers.

File: docs/root/intro/arch_overview/http/upgrades.rst
Comment (RyanTheOptimist) [inline docs/root/intro/arch_overview/http/upgrades.rst:55]: Nit: RFC 8841 defines Extended CONNECT for HTTP/2, not HTTP/1. The analogous HTTP/1 mechanism is Upgrade https://datatracker.ietf.org/doc/html/rfc7230#section-6.7

**Review by RyanTheOptimist — COMMENTED** [review]
Is this currently gated on a Quiche change to support the HTTP/3 setting?

**Comment by alyssawilk** [comment]
it's not gated on the settings given it works without them, but I added a TODO to follow through on that part.
@danzh2010 would you be willing to do the upstream work? I suspect you can do it more easily than I can :-/

**Comment by RyanTheOptimist** [comment]
> it's not gated on the settings given it works without them, but I added a TODO to follow through on that part.
> @danzh2010 would you be willing to do the upstream work? I suspect you can do it more easily than I can :-/

Ah, ok that SGTM. Presumably no clients will make use of this behavior without the setting being sent but no need to wait for that.

LGTM

**Review by RyanTheOptimist — DISMISSED** [review]

**Review by danzh2010 — COMMENTED** [review]
> it's not gated on the settings given it works without them, but I added a TODO to follow through on that part.
> @danzh2010 would you be willing to do the upstream work? I suspect you can do it more easily than I can :-/

Yeah, I can add the custom setting in QUICHE. Should QUICHE also reject the CONNECT request if the setting is not sent or leave it to Envoy? If the former, we might break some envoy deployment which uses CONNECT without receiving the special setting as how it works now.

File: api/envoy/config/core/v3/protocol.proto
Comment (RyanTheOptimist) [reply api/envoy/config/core/v3/protocol.proto:489]: Yes, the line below should be 8441 (websockets over HTTP/2) not 8841 (SCTP + Datagrams). Good catch!

File: source/common/http/header_utility.cc
Comment (RyanTheOptimist) [reply source/common/http/header_utility.cc:343]: Protocol and path MUST both be sent for extended CONNECT. So if one is sent but not the other, yes, I think that's invalid,

**Review by RyanTheOptimist — COMMENTED** [review]
> Yeah, I can add the custom setting in QUICHE. Should QUICHE also reject the CONNECT request if the setting is not sent or leave it to Envoy? If the former, we might break some envoy deployment which uses CONNECT without receiving the special setting as how it works now.

I think it would be fine to reject malformed CONNECT requests. When the setting is not present it's malformed to have path, scheme or protocol. When the setting is present it's malformed to have path or protocol absent. I *think* CONNECT + H3 must not yet be used in production so I *think* it won't cause problems. But Alyssa would know for sure.

File: test/integration/http_integration.cc
Comment (danzh2010) [reply test/integration/http_integration.cc:260]: How about this one?

**Comment by alyssawilk** [comment]
@mattklein123 I think this is ready for maintainer pass!

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:493]: I'm going to eventually convert this into an annotation. I assume that for a primitive type, if it's set to something other than default, I can determine that the annotation is set in that case but not if it's left to the default?

Basically I'm wondering if we need to make this a WKT vs. a primitive or not for fields. cc @envoyproxy/api-shepherds who know more about proto.

File: source/common/runtime/runtime_features.cc
Comment (mattklein123) [inline source/common/runtime/runtime_features.cc:97]: Should this be documented somewhere?

**Review by mattklein123 — COMMENTED** [review]
Thanks LGTM with some small API/doc questions.

/wait-any

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:493]: I'd think it'd be fine if for no other reason than setting it false is the same as not-present and doesn't need an alpha tag.  it's really only the behavior when this is explicitly set true that may change depending on the RFC

Comment (mattklein123) [reply api/envoy/config/core/v3/protocol.proto:493]: OK sounds good. I will see if this case works OK when I implement the feature.

**Review by mattklein123 — DISMISSED** [review]
Thanks!

File: api/envoy/config/core/v3/protocol.proto
Comment (RyanTheOptimist) [inline api/envoy/config/core/v3/protocol.proto:492]: Now that the IETF HTTP Working Group has adopted this document, the best link would now be 

https://datatracker.ietf.org/doc/draft-ietf-httpbis-h3-websockets/

Mind switching it?

File: api/envoy/config/core/v3/protocol.proto
Comment (htuch) [reply api/envoy/config/core/v3/protocol.proto:493]: Yeah, I think we do need in general to make WIP annotated things WKT/optional/messages when the default value is material, but if their absence is a no-op then we can skip.

File: test/integration/http_integration.cc
Comment (danzh2010) [reply test/integration/http_integration.cc:260]: I thought the ENABLE_CONNECT_PROTOCOL setting is supposed to be sent by server. So the the knob should only be used by server code. No?

File: source/common/http/header_utility.cc
Comment (DavidSchinazi) [inline source/common/http/header_utility.cc:343]: As per [the spec](https://datatracker.ietf.org/doc/html/rfc8441#section-4), when `:protocol` is included, both `:path` and `:scheme` are required. Should we validate both?

File: source/common/quic/envoy_quic_server_stream.cc
Comment (DavidSchinazi) [inline source/common/quic/envoy_quic_server_stream.cc:49]: FWIW I'm planning on adding support for this SETTING in QUICHE for Privacy Proxy, I plan on always sending it and validating the presence of pseudo-headers in QUICHE

File: test/integration/http_integration.cc
Comment (DavidSchinazi) [reply test/integration/http_integration.cc:260]: I'd imagine that on the client we'd use this option to fail the extended connect before it's sent? Or is it a no-op on the client?

File: api/envoy/config/core/v3/protocol.proto
Comment (DavidSchinazi) [inline api/envoy/config/core/v3/protocol.proto:494]: When we had discussed this last, I was under the impression that we wanted an HTTP/3 option to enable CONNECT altogether. Is that no longer the case? Are we allowing non-extended CONNECT by default?

**Review by DavidSchinazi — COMMENTED** [review]
Adding some random thoughts. Sorry I'm late to the party, extended CONNECT just became a dependency for privacy proxies and I saw comments on that fly by :)

File: source/common/quic/envoy_quic_server_stream.cc
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_server_stream.cc:49]: Thanks for taking this on. We should perhaps discuss this on the QUICHE PR/CL, but I'm not convinced that always sending the setting is the right choice. But let's discuss there.

File: api/envoy/config/core/v3/protocol.proto
Comment (RyanTheOptimist) [reply api/envoy/config/core/v3/protocol.proto:494]: Alyssa can correct me, but if I'm remembering our conversation correctly, I believe vanilla CONNECT is controlled by an HCM knob and we did not need a per-protocol vanilla CONNECT knob. (The HTTP/2 option, while named "allow_connect" actually enables extended connect so this HTTP/3 is parallel to the HTTP/3 option but with a more accurate name)

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto]: done

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:494]: Yep, we've gone over that a couple of times now :-) Old-style CONNECT gets rejected in the HCM if there's not connect matchers, and so we only need to control extended connect in the protoco-specific options.  I tried to also make sure the documentation pretty clear on this point (https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/upgrades#connect-support) but please let me know if you have ideas of how I could improve it!

File: source/common/http/header_utility.cc
Comment (alyssawilk) [reply source/common/http/header_utility.cc:343]: :scheme is required for all non-CONNECT requests and is supposed to be done prior to this point as we expect core protocol invariant to be enforced by the codec (this is pre-RFC and extension so less so for this). If QUICHE isn't doing that enforcement, I'd argue it probably should so thanks for volunteering to pick it up below.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_stream.cc:49]: @DavidSchinazi Actually I'm already working on a QUICHE change to add this setting per previous discussion. I can take this feature if you don't mind?

Comment (DavidSchinazi) [reply source/common/quic/envoy_quic_server_stream.cc:49]: @danzh2010 that's fine by me! From Alyssa's other comment I think QUICHE should validate that the correct headers are present. Please add me as a reviewer on the change

File: api/envoy/config/core/v3/protocol.proto
Comment (DavidSchinazi) [reply api/envoy/config/core/v3/protocol.proto:494]: Thanks for clarifying!

**Comment by mattklein123** [comment]
Sorry needs a main merge.

/wait

**Comment by alyssawilk** [comment]
Normally it's all "ugh, merge conflict."
I am SO excited about this merge conflict :-) :-)

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
